Common/Securities/FutureOption/FuturesOptionsUnderlyingMapper.cs

```diff
@@ -0,0 +1,172 @@
+                { "ZT", (d, ld) => ContractMonthSerialLookupRule(Symbol.Create("ZT", SecurityType.Future, Market.CBOT), d, ld.Value) },
+                { "ZW", (d, ld) => ContractMonthSerialLookupRule(Symbol.Create("ZW", SecurityType.Future, Market.CBOT), d, ld.Value) },
+
+                // COMEX
+                { "HG", (d, _) => ContractMonthYearStartThreeMonthsThenEvenOddMonthsSkipRule(d, true) },
+                { "SI", (d, _) => ContractMonthYearStartThreeMonthsThenEvenOddMonthsSkipRule(d, true) },
+                { "GC", (d, _) => ContractMonthEvenOddMonth(d, false) }
+            };
+        }
+
+        /// <summary>
+        /// Gets the FOP's underlying Future. The underlying Future's contract month might not match
+        /// the contract month of the Future Option when providing CBOT or COMEX based FOPs contracts to this method.
+        /// </summary>
+        /// <param name="futureOptionTicker">Future option ticker</param>
+        /// <param name="market">Market of the Future Option</param>
+        /// <param name="futureOptionExpiration">Expiration date of the future option</param>
+        /// <param name="date">Date to search the future chain provider with. Optional, but required for CBOT based contracts</param>
+        /// <returns>Symbol if there is an underlying for the FOP, null if there's no underlying found for the Future Option</returns>
+        public Symbol GetUnderlyingFutureFromFutureOption(string futureOptionTicker, string market, DateTime futureOptionExpiration, DateTime? date = null)
+        {
+            var futureTicker = FuturesOptionsSymbolMappings.MapFromOption(futureOptionTicker);
+            var canonicalFuture = Symbol.Create(futureTicker, SecurityType.Future, market);
+            // Get the contract month of the FOP to use when searching for the underlying.
+            // If the FOP and Future share the same contract month, this is reused as the future's
+            // contract month so that we can resolve the Future's expiry.
+            var contractMonth = FuturesOptionsExpiryFunctions.GetFutureContractMonth(canonicalFuture, futureOptionExpiration);
+
+            if (_underlyingFuturesOptionsRules.ContainsKey(futureTicker))
+            {
+                // The provided ticker follows some sort of rule. Let's figure out the underlying's contract month.
+                var newFutureContractMonth = _underlyingFuturesOptionsRules[futureTicker](contractMonth, date);
+                if (newFutureContractMonth == null)
+                {
+                    // This will only happen when we search the Futures chain for a given contract and no
+                    // closest match could be made, i.e. there are no futures in the chain that come after the FOP's
+                    // contract month.
+                    return null;
+                }
+
+                contractMonth = newFutureContractMonth.Value;
```

Comment:
So this means that the response of `FuturesOptionsExpiryFunctions.GetFutureContractMonth` isn't always correct, what about the other usages, like IB, I think they shouldn't be directly calling it? but instead calling this new mapper class. 

~We should add a big comment at `FuturesOptionsExpiryFunctions.GetFutureContractMonth` about this and probably point to this class~. Or actually even **better** we should have a single method that returns the correct future contract month directly 🤔 else it's like a bug waiting to happen.

Might we worth reviewing if there are any other similar cases where these helper methods are returning a buggy response which requires the caller to know about it and adjust correctly with another call